### ISSUE TITLE
Support `concat()` or similar?

### ISSUE THREAD

**Issue by polarathene** [polazarus/hipstr#20]
https://docs.rs/hipstr/0.4.0/hipstr/bytes/struct.HipByt.html#method.concat

If I have a `[&str; 5]` this would use 80 bytes to represent the string data correct? But with a `HipStr` if the actual data were 23 bytes or less I could compose that representation as a single `HipStr` and avoid allocating a new `String`?

I'm not sure if that would be any different than just creating a new `HipStr` and doing `push_str()`?:

```rs
let iter = ["key".as_bytes(), "value".as_bytes()];
iter.fold(HipByt::new(), |mut s, v| { s.push_slice(v); s })

let iter = ["key", "value"].into_iter();
iter.fold(HipStr::new(), |mut s, v| { s.push_str(v); s })
```

I've seen various methods for `HipStr` and `HipByt` that will output their type from a slice of the existing data, but not much for constructing the type from multiples of the type (_or `&str` / `&[u8]`_).

```rs
let parts = [key, "=\"", value, "\""];

// key="value"
let formatted = parts.concat();
```

Would it be worthwhile for `HipStr` and `HipByt` to support `concat()` to create an inline value more easily from multiple values?

---

In my case, I have a HashMap collection of KV pairs that I iterate through to create `key="value"` sequences, then the iterator composes those into a single `String` output.

That is fast with references (`&str` or `&HipStr`), and from what I understand this isn't something `HipStr` could improve on? So constructing a `HipStr` from an array of `&str` / `&HipStr` doesn't appear helpful due to the extra allocations/operations required?

**Comment by polarathene** [comment]
**TL;DR:** My current use-case isn't likely to benefit from these special string types it seems 😓 

Still, the feature request might warrant the convenience of deriving a `HipStr` / `HipByt` from multiple values which methods like `concat()` seem appropriate for.

---

I was curious and adapted `HipStr` and `HipByt` to be compatible with the `Extend` trait. Below is an example with a wrapper type to extend an iterator into a `String`:

```rs
// Wrapper struct for `impl Extend`, since we cannot impl for `String`
struct StringWrapper(String);

impl StringWrapper {
    fn with_capacity(size: usize) -> Self {
        Self(String::with_capacity(size))
    }
    fn push(&mut self, value: char) {
        self.0.push(value);
    }
    fn push_str(&mut self, value: &str) {
        self.0.push_str(value);
    }
}

impl From<MyString> for String {
    fn from(s: MyString) -> String {
        s.0
    }
}

// An iterator over `[&str; 5]` or `[&HipStr; 5]`:
impl<'a, StrRef> Extend<[&'a StrRef; 5]> for StringWrapper
where StrRef: AsRef<str> + ?Sized {
    fn extend<T: IntoIterator<Item=[&'a StrRef; 5]>>(&mut self, iter: T) {
        for item in iter {
            item.into_iter().for_each(|s| {
                self.0.push_str(s.as_ref());
            })
        }
    }
}

// For a flattened iterator over `&str` / `&HipStr` items:
impl<'a, StrRef> Extend<&'a StrRef> for StringWrapper 
where StrRef: AsRef<str> + ?Sized {
    fn extend<T: IntoIterator<Item=&'a StrRef>>(&mut self, iter: T) {
        iter.into_iter().for_each(|s| {
                self.0.push_str(s.as_ref());
        })
    }
}
```

I also tried with `HipByt`, and likewise (_if just pushing the references to the `String` as-is_) it's quick too 😄 

Only when trying to merge into a single `HipByt` was it quite slow (5x); even when the content remained inlined throughout:

```rs
// HipByt equivalent of above extend, but also generic:
// HipStr internally uses HipByt (which internally uses Raw), might be better?
struct HipBytWrapper<'a, const N: usize>([&'a HipByt<'a>; N]);

impl<'a, const N: usize> Extend<HipBytWrapper<'a, N>> for StringWrapper {
    fn extend<T: IntoIterator<Item=HipBytWrapper<'a, N>>>(&mut self, iter: T) {
        for item in iter {
            // Convert each HipByt to a HipStr to push to the inner String
            item.0.into_iter().for_each(|hbyt| {
                let s = unsafe { HipStr::from_utf8_unchecked(hbyt.clone()) };

                assert!(s.len()<24);
                assert!(!s.is_allocated());
                // Requires `into_owned()` (2x slower)
                // Due to static borrows, otherwise this would be true without `into_owned()`
                // assert!(s.is_inline());
                self.0.push_str(&s);
            })

            // 5x slower:
            /*
            // Combine all HipByt into single HipByt (remains inline):
            let b = item.0.into_iter().fold(HipByt::new(), |mut new, hybt| {
                new.push_slice(hybt);
                new
            });
            // assert!(b.is_inline());
            // Single `push_str()` per key/value pair:
            let s = unsafe { HipStr::from_utf8_unchecked(b) };
            self.0.push_str(&s);
            */
        }
    }
}
```

My actual processing is a bit more involved but for this topic, it's roughly this:

```rs
type MapString = String;
type Params = indexmap::IndexMap<MapString, MapString>;
/* HipStr types:
type MapString<'a> = hipstr::HipStr<'a>;
type Params<'a> = indexmap::IndexMap<MapString<'a>, MapString<'a>>;
*/

pub fn param_into_string(params: &Params) -> String {
    let mut s = StringWrapper::with_capacity(64);

    // Prepare iterator
    let iter = params.into_iter().map(|(key, value)| {
        let kv_pair: [&str; 5] = [" ", key, "=\"", value, "\""];

        /*
        // HipStr is equivalent perf to &str when KV values are of the same type:
        let kv_pair: [&HipStr; 5] = [SPACE, key, EQUALS, value, END];

        // HipByt is slower to get an &str / HipStr from, requires wrapper for `extend()`:
        let kv_pair: [&HipByt; 5] = [BSPACE, key, BEQUALS, value, BEND];
        HipBytWrapper(kv_pair)
        */

        kv_pair
    });

    // Append all pairs to the output String
    s.extend(iter);
    s.into()
}

#[divan::bench]
fn bench_params(b: Bencher) {
    b.with_inputs(|| {
        [
            ("key A", "value A"),
            ("key B", "value B"),
            ("key C", "value C"),
            ("key D", "value D"),
        ].into_iter()
        .map(|(k,v)| (MapString::from(k), MapString::from(v)))
        .collect::<Params>()
    }).bench_refs(|params| {
        param_into_string(params);
    });
}
```

**Comment by polazarus** [comment]
Sorry it seems I missed your issue.

Mmm I think something could be done to make a fast `concat`.
I am not sure about the API though... `Concat` is experimental.
Maybe just a new constructor on Hip*? e.g.

```rust
HipStr::concat(slices: &[impl AsRef<str>])
```

I am not sure about `extend` though. IMO you need specialization to make that fast when the iterator is `Copy` for instance.

**Comment by polazarus** [comment]
Note to self, `join` too.

**Comment by polazarus** [comment]
I think your test is skewed though, you will need some kind of  `push_slice_unchecked` to make it fast (with some strong precondition on the stability of the underlying representation).

**Comment by polarathene** [comment]
I've not got much to contribute to the discussion here beyond what I was seeking to achieve with `HipStr` / `HipByt` as a faster way to construct the `key="value"` string from HashMap entries.

I'm not sure if that's viable, it's been a while since I experimented with different approaches and crates but I think this one had the most promise. I might get around to revisiting that project and putting the bench on Github at some point 😅 

---

> I think your test is skewed though

Oh maybe! This was my first time benching with Rust.

I do recall [encountering some odd issues](https://github.com/polarathene/reproduction-rust-bench-regression) (_changes I wouldn't expect the compiler to mess with bench times_).

> you will need some kind of `push_slice_unchecked` to make it fast (with some strong precondition on the stability of the underlying representation).

Is the stability in this case not a given with the usage of `HipByt` to `HipStr`? I had  5 elements to compose where 3 are static values:

```rs
// HipByt or HipStr:
let kv_pair: [&HipStr; 5] = [SPACE, key, EQUALS, value, END];
```

`key` and `value` being input values from the HashMap that I'm converting into a single long `String` of concatenated `kv_pair`s. `push_slice_unchecked()` would be only relevant to the construction of a single `HipStr` / `HipByt` from those elements right? (_which may exceed the inline 23 bytes for a `kv_pair` depending on `key` + `value` length_).

---

I recall `fold()` was the faster approach I had benched (_I don't think any of SSO types I had tried helped due to allocation?_). At a quick glance on the project I think this was my best attempt:

```rs
// Converts a map of IDs with key/value pairs (HashMap<String, HashMap<String, String>>) to this output:
// [id_1 key_a="value a" key_b="value b" key_c="value c"][id_42 key_x="value x" key_y="value y" key_z="value z"]
pub fn fold(data: &Map) -> String {
    let s = String::with_capacity(data.count_bytes());

    let iter = data.iter().map(|(id, params)| {
        let kv_pairs = params.iter().flat_map(|(key, value)| {
            [" ", key, "=\"", value, "\""]
        });

        (id.as_str(), kv_pairs)
    });

    iter.fold(s, |mut s, (id, kv_pairs)| {
        s.push('[');
        s.push_str(id);
        s.extend(kv_pairs); // 50ns (approx 50% of time used here, other pushes only approx 6ns)
        s.push(']');

        s
    })
}
```

If `s.extend(kv_pairs)` was where the bulk of time was spent (_the comment claiming that would have been a bit naive in approach_), that's probably not something `hipstr` can do much about to improve since it's allocating to a larger `String` (_for serialization output_)?

---

I wouldn't worry about this too much 👍 

There doesn't seem to be any new interest in the feature request since I raised it from other users, and I did note previously that since it isn't likely to improve my use-case, it really amounts to a construction convenience method:

> _Still, the feature request might warrant the convenience of deriving a `HipStr` / `HipByt` from multiple values which methods like `concat()` seem appropriate for._

> Maybe just a new constructor on Hip*? e.g.
> ```rs
> HipStr::concat(slices: &[impl AsRef<str>])
> ```

Sounds good! 😎
